### PR TITLE
ScriptProcessorNode

### LINKED ISSUE #258: Add deprecated ScriptProcessorNode
Constructor: https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-createscriptprocessor
`createScriptProcessor(bufferSize, numberOfInputChannels, numberOfOutputChannels)`

Interface:
```
[Exposed=Window]
interface ScriptProcessorNode : AudioNode {
    attribute EventHandler onaudioprocess;
    readonly attribute long bufferSize;
};
```

Event with the input/output buffers:
```
[Exposed=Window]
interface AudioProcessingEvent : Event {
    constructor
 (DOMString type
, AudioProcessingEventInit eventInitDict
);
    readonly attribute double playbackTime;
    readonly attribute AudioBuffer inputBuffer;
    readonly attribute AudioBuffer outputBuffer;
};
```

I presume these `onaudioprocess` callbacks should run in the control thread. So an implementation would be:
- in the renderer, buffer samples up to the required bufferSize
- convert to control side AudioBuffer
- dispatch event
- when the event is handled, ship output back to render thread
- play output samples untill output buffer is exhausted
- repeat..

### PR DESCRIPTION
**Pull request by orottier**
Big TODO still: it does not work with the OfflineAudioContext because it does not run an event loop.
And if it would, it would not function properly because the rendering might run faster than the event loop can provide new data.

Fixes #258

### CODE REVIEW

**Comment by b-ma** [comment]
Cool, looks like a really nice start! 

I had some ideas I wanted to test to hopefully avoid allocations but given my structural lack of time these days this will wait for further refactoring :)

> Big TODO still: it does not work with the OfflineAudioContext because it does not run an event loop.

Just to say, it looks like another argument (with https://github.com/ircam-ismm/node-web-audio-api/pull/95) for having an event loop on the `OfflineAudioContext`... with the same, agreed, limitation of usability / usefulness :)

**Comment by orottier** [comment]
> I had some ideas I wanted to test to hopefully avoid allocations but given my structural lack of time these days this will wait for further refactoring :)

That sounds doable! I don't consider the real-time safety a blocker for merging this version though, as the ScriptProcessorNode is inherently not real-time safe. Our event loop (like the JS one) is not multithreaded/coorporative

> 
> > Big TODO still: it does not work with the OfflineAudioContext because it does not run an event loop.
> 
> Just to say, it looks like another argument (with [ircam-ismm/node-web-audio-api#95](https://github.com/ircam-ismm/node-web-audio-api/pull/95)) for having an event loop on the `OfflineAudioContext`... with the same, agreed, limitation of usability / usefulness :)

I figured you would say that :) But it will need some experimenting to see if adding the event loop to OfflineAudioContext is enough for all uses cases to work. If the render thread requires the data before the event loop can provide it, what should the behaviour be? The spec does not provide information on this. 

Using our EventLoop, we cannot block in the ScriptProcessorRenderer so we emit silence if the buffer is not provided in time. This is problematic if the ScriptProcessor is doing heavy lifting, and the rest of the audio context is very light. I guess that is not acceptable for OfflineAudioContext so we need a different solution - further complicated by the fact that `start_rendering_sync` may end up in a deadlock if the render thread blocks on the control thread (they are the same in that case!)

**Comment by b-ma** [comment]
> If the render thread requires the data before the event loop can provide it, what should the behaviour be? The spec does not provide information on this.
> Using our EventLoop, we cannot block in the ScriptProcessorRenderer so we emit silence if the buffer is not provided in time.

In my opinion this seems perfectly fine, but just to make sure I have made a small test-bed app to check browsers behaviour (https://github.com/b-ma/test-script-processor-node), and... 
- Chrome does seem to somehow wait for the script events in the calculation of the offline context, you can see in the logs that the audio signal starts at index 512 as expected (actually 513 but because first sample of sine is zero)
- Firefox doesn't

(removing the `fib(1000000);` in audio process callback, doesn't change much the result)

So I guess it is up to us to define what behaviour we want to provide

**Comment by orottier** [comment]
>     * Chrome does seem to somehow wait for the script events in the calculation of the offline context, you can see in the logs that the audio signal starts at index 512 as expected (actually 513 but because first sample of sine is zero)
>     * Firefox doesn't
> So I guess it is up to us to define what behaviour we want to provide

That's very helpful. I will experiment with bringing the EventLoop into the OfflineAudioContext then and we will use the Firefox approach for now.
I can probably mitigate most of the performance impact by running an EventLoop per executable instead of per (Offline)AudioContext. I will require some rework though

**Comment by orottier** [comment]
@b-ma any final comments? With the offline event loop fixed, we are all set here now

**Comment by b-ma** [comment]
Appart my previous comments (in particular the first one about introducing a `ScriptProcessorOptions` for consistency with other nodes), looks good to me

**Comment by orottier** [comment]
> ScriptProcessorOptions

I don't see any comment regarding this, perhaps they are still lingering in a pending review?

**Comment by b-ma** [comment]
Damn, didn't get there was another step

File: src/node/script_processor.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        self.channel_config.set_count(count, self.registration());
+    }
+}
+
+impl ScriptProcessorNode {
+    pub(crate) fn new<C: BaseAudioContext>(
+        context: &C,
+        buffer_size: usize,
+        number_of_input_channels: usize,
+        number_of_output_channels: usize,
```
[the remark this answers is not in the document]
Comment (orottier) [reply src/node/script_processor.rs:63]: Done, and marked it `pub`

File: src/node/script_processor.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            self.next_output_buffer.clear();
+            let silence = output.clone();
+            self.next_output_buffer.resize(number_of_quanta, silence);
+        }
+
+        if !self.output_buffer.is_empty() {
+            *output = self.output_buffer.remove(0);
+        }
+
+        true // TODO - spec says false but that seems weird
```
[the remark this answers is not in the document]
Comment (orottier) [reply src/node/script_processor.rs]: Yeah, I realize there is no need to deviate anyway. Fixed

File: src/node/script_processor.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                    .zip(input_samples.iter_mut())
+                    .for_each(|(c, o)| {
+                        o[offset..(offset + RENDER_QUANTUM_SIZE)].copy_from_slice(c);
+                    });
+            });
+            let input_buffer = AudioBuffer::from(input_samples, scope.sample_rate);
+
+            // create a suitable output AudioBuffer
+            let output_samples = vec![vec![0.; self.buffer_size]; self.number_of_output_channels];
+            let output_buffer = AudioBuffer::from(output_samples, scope.sample_rate);
```
[the remark this answers is not in the document]
Comment (orottier) [reply src/node/script_processor.rs:231]: That is not simple because we need a new buffer every few render quanta. I will write a separate issue for this

**Pull request merged**